### PR TITLE
Add some warnings to std::env::current_exe

### PR DESCRIPTION
**Pull request by steveklabnik**
/cc #21889 @rust-lang/libs @semarie 

I started writing this up. I'm not sure if we want to go into other things and in what depth; we don't currently have a lot of security-specific documentation to model after.

Thoughts?

### CODE REVIEW

**Comment by brson** [comment]
lgtm

File: src/libstd/env.rs [hunk 1]
```diff
@@ -493,6 +493,21 @@ pub fn temp_dir() -> PathBuf {
 /// that can fail for a good number of reasons. Some errors can include, but not
 /// be limited to, filesystem operations failing or general syscall failures.
 ///
+/// # Security
+///
+/// This function should be used with care, as its incorrect usage can cause
+/// security problems. Specifically, as with many operations invovling files and
```
Comment (alexcrichton) [inline src/libstd/env.rs]: "invovling"

**Comment by alexcrichton** [comment]
Thanks @steveklabnik! Perhaps we could emphasize that calling this function itself only has security implications if you assume some property about the return value which may have security implications? I don't find the example of overwriting something too compelling because if you can overwrite the executable then you can, well, overwrite the executable (as in, `current_exe` is the least of your privileges).

The real problem here is you can escalate privileges, which I found [this example](http://securityvulns.com/Wdocument183.html) to be the most compelling. You can trivially cause `std::env::current_exe()` to return something arbitrary, for example:

```
$ echo 'fn main() { println!("{:?}", std::env::current_exe()); }' > foo.rs
$ rustc foo.rs
$ ./foo
Ok("/home/alex/foo")
$ ln foo bar
$ ./bar
Ok("/home/alex/bar")
```

For example I would consider the compiler's usage of this function "not insecure" because although you could cause rustc to link against a random libstd, it's already generating arbitrary code... The problem in that linked post was that the path would be used for executing an executable as root, which is clearly going to end badly if you can run arbitrary executables.

**Comment by semarie** [comment]
@alexcrichton if you call the function, it is for the returned pathname (you don't call it just for a side effect): so you always assume one property: you will get the pathname of the running executable.

One problem is the moving definition of the returned pathname against the platform.
It could be:
  -  the pathname used at `execve(2)` time (the pathname doesn't change if the file is moved)
  - a best effort for having the pathname of the running executable (one OS which store it in some cache system, and under memory pressure could be flushed - or some other failure conditions)
  - an error because the file was deleted or moved
  - an error because the platform don't support retrieving this information
  - or always the good pathname (even if moved) or an error if removed (the better implementation)

So basically, if you got a pathname, you couldn't assume the pathname **will** be the right one.

@steveklabnik I think your description is good enough. Maybe just add a note for saying that the accuracy of the returned pathname is platform dependant (and so should be assumed to be wrong) ?

**Comment by semarie** [comment]
so, no documentation warning at all ?

**Comment by steveklabnik** [comment]
Whoops. I was looking at my "open PRs" and saw only the one I just sent. I then cleaned up a _lot_ of my old branches. I guess this PR was still outstanding? Will re-send later today.

**Comment by alexcrichton** [comment]
@steveklabnik  thoughts on updating with some of the discussion above?

**Comment by steveklabnik** [comment]
I will get on it today.

**Comment by steveklabnik** [comment]
Updated, to use the more real example, and link to the vulnerability. I do think that it's stronger as an example. Still not _totally_ happy with the wording here, but it might be Good Enough.

**Comment by alexcrichton** [comment]
@bors: r+ c4730daf450bafe51e76d430f94d2c59c814e53c

Looks good to me!

**Comment by alexcrichton** [comment]
@bors: retry

* #34595

**Comment by alexcrichton** [comment]
@bors: retry

* #34595

**Comment by semarie** [comment]
thanks

**Pull request merged**